src/unix/stream.c

```diff
@@ -826,16 +841,22 @@ static void uv__write(uv_stream_t* stream) {
     do {
       n = sendmsg(uv__stream_fd(stream), &msg, 0);
     }
-#if defined(__APPLE__)
-    /*
-     * Due to a possible kernel bug at least in OS X 10.10 "Yosemite",
-     * EPROTOTYPE can be returned while trying to write to a socket that is
-     * shutting down. If we retry the write, we should get the expected EPIPE
-     * instead.
-     */
-    while (n == -1 && (errno == EINTR || errno == EPROTOTYPE));
-#else
-    while (n == -1 && errno == EINTR);
+    while (n == -1 && ok_errno);
+#ifdef TCP_FASTOPEN
+  } else if (stream->flags & UV_TCP_FASTOPEN) {
+    struct sockaddr *addr = (struct sockaddr *)&((uv_tcp_t *)stream)->addr;
+    socklen_t addrlen = 0;
+    switch (addr->sa_family) {
+    case AF_INET: addrlen = sizeof(struct sockaddr_in); break;
+    case AF_INET6: addrlen = sizeof(struct sockaddr_in6); break;
+    default: err = -EINVAL; goto error;
+    }
+    do {
+      n = sendto(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len,
+		 MSG_FASTOPEN, addr, addrlen);
+    }
+    while (n == -1 && (ok_errno || errno == EINPROGRESS));
+    stream->flags &= ~UV_TCP_FASTOPEN;
```

Comment:
The TFO applies only for the first message in the stream (as it sends the data along the handshake), for the subsequent packets, it doesn't make any difference and it's better to use the standard way of sending packets.

Reply:
The documentation on TCP FastOpen isn't the greatest, but I found this: https://dnsprivacy.org/wiki/display/DP/TCP+Fast+Open#TCPFastOpen-LinuxAPI

- On the client, the call to connect() is replaced with a call to sendmsg() or sendto() with the flags parameter set to MSG_FASTOPEN. 
  - For blocking sockets this performs both the handshake and sends the data before returning.

  - For non-blocking socket it returns the number of bytes buffered and sent in the SYN packet. 

    - If the cookie is available it will send the data before returning.

    - If the cookie is not available locally, it returns -1 with errno EINPROGRESS, and sends a SYN with TFO cookie request automatically. The caller needs to write the data again when the socket is connected with a call to send() or similar.

- Subsequent writes on this socket must also be done with send() or similar. A subsequent call with sendto() will return the error EISCONN.